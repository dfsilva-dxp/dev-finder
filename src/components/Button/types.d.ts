import { ButtonHTMLAttributes, ReactNode } from "react";

export interface IButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  icon?: ReactNode;
  children?: ReactNode;
}
