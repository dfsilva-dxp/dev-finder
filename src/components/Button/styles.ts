import styled, { css } from "styled-components";

export const Wrapper = styled.button`
  ${({ theme }) => css`
    background-color: transparent;
    border: 0;
    border-radius: ${theme.border.radius};
    box-shadow: none;
    font-size: ${theme.font.sizes.xsmall};
    font-weight: ${theme.font.weight.medium};
    color: ${theme.colors.text};
    text-transform: uppercase;

    display: flex;
    align-items: center;
    justify-content: center;
    gap: ${theme.spacings.xxsmall};

    &:hover,
    &:focus {
      outline: 2px solid ${theme.colors.primary};
      outline-offset: ${theme.border.outlineOffiset};
    }
  `}
`;
