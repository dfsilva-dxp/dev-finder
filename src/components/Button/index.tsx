import { Wrapper } from "./styles";
import { IButtonProps } from "./types";

export default function Button({
  icon,
  children = "Button",
  ...props
}: IButtonProps) {
  return (
    <Wrapper {...props}>
      {children}
      {!!icon && icon}
    </Wrapper>
  );
}
