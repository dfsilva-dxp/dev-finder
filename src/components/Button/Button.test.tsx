import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SunDim } from "phosphor-react";
import { act } from "react-dom/test-utils";
import { renderWithTheme } from "utils/test/helpers";

import Button from ".";

describe("<Button/>", () => {
  it("should render currectly", () => {
    renderWithTheme(<Button>Theme</Button>);

    expect(screen.getByRole("button", { name: /theme/i })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /theme/i })).toHaveStyle({
      "background-color": "transparent",
      border: "0",
      "font-size": "0.75rem"
    });
    expect(screen.getByRole("button", { name: /theme/i })).toHaveStyleRule(
      "outline",
      "2px solid #FE5873",
      {
        modifier: ":hover"
      }
    );
  });

  it("should render an icon", () => {
    const { container } = renderWithTheme(
      <Button icon={<SunDim size={18} data-testid="icon" />}>Theme</Button>
    );

    expect(screen.getByRole("button", { name: /theme/i })).toBeInTheDocument();
    expect(screen.getByTestId("icon")).toBeInTheDocument();

    expect(container.parentElement).toMatchSnapshot();
  });

  it("should be accessible with tab", () => {
    renderWithTheme(<Button>Theme</Button>);

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(screen.getByRole("button", { name: /theme/i })).toHaveFocus();
  });

  it("should be called the fn() when button is cliecked", async () => {
    const toggleIcon = jest.fn();

    act(() => {
      renderWithTheme(<Button onClick={toggleIcon}>Theme</Button>);
    });

    const button = screen.getByRole("button", { name: /theme/i });

    act(() => {
      button.dispatchEvent(new Event("click", { bubbles: true }));
    });

    expect(toggleIcon).toHaveBeenCalledTimes(1);
  });
});
