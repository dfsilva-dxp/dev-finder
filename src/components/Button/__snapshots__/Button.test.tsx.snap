// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should render an icon 1`] = `
.c0 {
  background-color: transparent;
  border: 0;
  border-radius: 0.25rem;
  box-shadow: none;
  font-size: 0.75rem;
  font-weight: 500;
  color: #FFFFFF;
  text-transform: uppercase;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 0.5rem;
}

.c0:hover,
.c0:focus {
  outline: 2px solid #FE5873;
  outline-offset: 0.125rem;
}

<body>
  <div>
    <button
      class="c0"
    >
      Theme
      <svg
        data-testid="icon"
        fill="currentColor"
        height="18"
        viewBox="0 0 256 256"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="128"
          cy="128"
          fill="none"
          r="60"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="128"
          x2="128"
          y1="36"
          y2="28"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="62.9"
          x2="57.3"
          y1="62.9"
          y2="57.3"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="36"
          x2="28"
          y1="128"
          y2="128"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="62.9"
          x2="57.3"
          y1="193.1"
          y2="198.7"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="128"
          x2="128"
          y1="220"
          y2="228"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="193.1"
          x2="198.7"
          y1="193.1"
          y2="198.7"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="220"
          x2="228"
          y1="128"
          y2="128"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="193.1"
          x2="198.7"
          y1="62.9"
          y2="57.3"
        />
      </svg>
    </button>
  </div>
</body>
`;
